### ISSUE TITLE
QR: Link preview to icons page if no browse; move random into drop-down and spawn re-roll button

### ISSUE THREAD

**Issue by nfagerlund** [dreamwidth/dw-free#2522]
https://dw-maintenance.dreamwidth.org/78131.html?thread=2943539#cmt2943539

- Put it on the icon preview whenever we don't turn it into a browse button? 
    - But then paid users don't have the link. 
    - But do they even need it if they have the browser? (Assume we're going to fix #2411 at some point in the near future.)
- Put it somewhere else? 
- Leave it out, bc there are other ways to get there and we don't need to add links that navigate you away from the comment form?

**Comment by rahaeli** [comment]
My vote goes to "leave it out because there are other ways to get there and adding the user tag somewhere will make /icons one click away anyway"

**Comment by Alyndra** [comment]
I was making pro/con lists in my head for a couple options here even before this issue opened but...you make a very good point, once I realized from seeing this comment and running to check that the icon in the standard hoverbox was, in fact, clickable. Maybe it needs a "hover over the icon and it says "ICONS" where the other hover says "BROWSE," haha? I wouldn't have realized it was clickable without looking at the bottom of my web browser for an address.

**Comment by rahaeli** [comment]
Ha, I was actually thinking of "click on the userhead icon to get to the profile, click on the icon link to get to the profile page" as the one click. Is there another way?

**Comment by Alyndra** [comment]
If you hover on the userhead icon it has a short list of links and the user's default icon, and the icon is clickable (I JUST DISCOVERED A THING)

**Comment by nfagerlund** [comment]
Yeah, that’s how I always get to my icons page. I like the idea of doing a “view all” mask that’s like the “browse” mask, tho!

**Comment by rahaeli** [comment]
Ohhh, yeah, the contextual hover popup! (I have that disabled because I hate things flying around just because I moused over something, so I didn't think of it.)

**Comment by Alyndra** [comment]
Thought about this some more. I am prepared to make a case that clicking on the preview icon should do _something_ icon-related instead of nothing, since for new or infrequent users who want to look at all their icons, it's much more intuitive to click the icon than to go the janky route of the userhead icon/profile.

At this point I wondered what current non-paid account usage looked like, so I made an account I'd been meaning to create anyway and played around with it. HIGHLY SCIENTIFIC RESULTS:

When a free account has no icons uploaded yet, the browse, random, and dropdown select boxes all vanish. Good. There is a site-standard icon preview, and clicking it does nothing. Not good. Ideal case, it should link to Manage Icons page; linking to /icons page would also be fine.

When a free account has 1 icon uploaded, the random and dropdown select pop back in. I'd get rid of Random here since all it does is change from default to your one icon. It's tempting to get rid of the dropdown, too, but users might want to differentiate between default (will change if they later change their default icon) and Icon 1 (will permanently stay that icon as long as you have it available no matter what your default is), so it probably ought to stay.

Do paid accounts with 0 or 1 icons also have to put up with the Browse and Random buttons? Likely. The utility of those buttons almost certainly begins when 2+ icons are uploaded, and otherwise they just take up space. No big deal on large screens, but it may become an issue when the QuickReply gets very narrow and has to collapse everything.

For no default set: Excitingly, when I hover over the icon (whether the dropdown is set to Default or Icon 1) there is a tooltip that says "Manage Icons." Crushingly, when I click, nothing happens. Was this already bug-logged? I think it was.

If the user has an icon set as default, then hovering over it will summon the hoverbox (lightbox?) with links to journal, profile, and icon page. This works on mobile and on desktop. So far, so good! HOWEVER, if I go into Settings, Display, "Show the contextual hover menu when you hover over" and choose either "nothing" or "userhead graphic only" THEN it is no longer clickable. I think the icon should be clickable in this context regardless of hover menu setting. Clicking the icon of the comment you are replying to brings you to that user's icon page, and this occurs all over DW, so for the icon preview not to link to your icon page in the quickreply form is subpar. 

Tl;dr It would be great if QR icon preview was clickable to /icons or /manage/icons?authas=username when 
A. No icons are uploaded
B. No default is set
C. Icon Hover Menu is disabled 
D. Unforeseen gremlins prevent hover menu usage (???)

It would additionally be great if it identified itself as clickable by text such as "VIEW ALL ICONS" when hovered over. 

Bonus tweak for optimal design (low priority): take out Random and Browse buttons when only 1 icon is uploaded. 

Multiple users have requested to be allowed to remove Browse and/or Random if unused, which might not be a bad idea?

...Eeep, sorry this got so long! Hitting Comment before it _grows..._

**Comment by rahaeli** [comment]
> Multiple users have requested to be allowed to remove Browse and/or Random if unused, which might not be a bad idea?

Nah, that would need an option, and "extra button that you know you aren't gonna use" isn't worth an option IMO.

**Comment by nfagerlund** [comment]
I think making the preview icon a link when browse is disabled is probably reasonable! Certainly easy to do, and non-disruptive for paid users, and would work like every other icon, like you mentioned. 

@Alyndra The more I think about your idea to make "random" an item in the dropdown menu instead of a button (wait, that was you, right? It feels like one of your ideas), the more I like it. Could even include an emoji 🔀 in the text for it, ha. Put it right under (default). 

![Screen Shot 2019-06-25 at 11 59 19 AM](https://user-images.githubusercontent.com/484309/60125501-c593a700-9740-11e9-99f5-d555755105e2.png)

Right now, clicking the random button changes the value of the select, and excludes "(default)" from the roulette. Easy enough to just trigger the exact same action when you select "random" instead, and exclude both "random" and "(default)" from the roulette.

**Comment by rahaeli** [comment]
Ohhhhhhh I love that

**Comment by Alyndra** [comment]
LOL experimenting with paid account, no icons uploaded (thanks @rahaeli ): Browse button, random button, and dropdown all disappear, as expected. Hovering over the site icon preview shows me the word 'Browse' but clicking on it does nothing, and no pop-up appears. Should also link to manage icons page, perhaps?

Haha got my icon back and started testing out the Browse pop-up on my little iPhone. It really doesn't work effectively. Even with only one icon uploaded it doesn't show the image, just the keywords, which overlap the 'select' button. And you can't tell which customizable options are selected because they all look the same. Is this a style issue, maybe? On Practicality now.

YES, it is. I applied a different style (Mobility) and got a wildly different looking Browse box. Hoo boy, this is fun. The second browse box has vast white space for dozens of icons despite my only having one.

...This is probably _anti-productive_ so I'm going to quit before figuring out how deep the rabbit hole goes.

**Comment by nfagerlund** [comment]
@Alyndra https://github.com/dreamwidth/dw-free/issues/2411

**Comment by nfagerlund** [comment]
That paid-but-no-icons behavior is WACKY. Why would the browse button be gone, but the preview button be messed up? Huh.

**Comment by nfagerlund** [comment]
OK, I've gotten "Random in the menu" working in a branch! 

- It goes right below (default). 
- It only appears if JS is enabled. (only matters for the talkform.)
- It uses `""` (empty string) as its value so it doesn't conflict with any real icon names, and if you somehow manage to submit the form with it selected (you'd have to disable JS _after_ summoning the quickreply), it'll act like the (default) option. 
- It works as expected (once I fixed the off-by one error that made the final icon never get picked). 

![random](https://user-images.githubusercontent.com/484309/60357933-cbc89400-9989-11e9-9537-8ed565ffc7c6.gif)

Right? So here's the only wrinkle: _it's a pain in the butt to repeatedly re-roll if you didn't like the result._ Especially if your icon list is a hundred deep. 

Do we still want to do this? @rahaeli @Alyndra

**Comment by Alyndra** [comment]
How would I select it if I am just using a keyboard and no mouse? Maybe do a keyboard shortcut for it that you can show in parenthesis on the Random line? (Sorry, spitballing…)

**Comment by nfagerlund** [comment]
@Alyndra It triggers on the "change" event, which fires when you do \~whatever\~ to confirm a pick from the menu! So with a keyboard, you'd tab to the menu, use arrow keys to pick random, and hit enter, and then it would randomize for you.

**Comment by Alyndra** [comment]
I guess what I'm asking is, if I tab to my dropdown now, I can hit p and it will take me to the first keyword starting with p; but I don't have a key for 🔀, so is the up/down arrow my only option?

**Comment by nfagerlund** [comment]
oh!!! I see what you mean. Yeah, the way I did it, up/down is your only option. Which indicates that I did it the wrong way!! I guess it should be the other way around, `(random) 🔀`.

**Comment by Alyndra** [comment]
So that might make it easier to re-roll, but I have no idea where the hassle threshold is, really.

**Comment by nfagerlund** [comment]
Hmm, I suppose once you choose "(random)" once, I could spawn a "Re-roll" button. That way there's a button, but it only takes up space once you've specifically indicated that you're That Kind of User.

**Comment by Alyndra** [comment]
...psst does (random) leave the dropdown if only one icon's uploaded?

**Comment by nfagerlund** [comment]
Not as of now, but no reason it couldn’t.

**Comment by nfagerlund** [comment]
claiming

**Comment by nfagerlund** [comment]
Capturing this here, because it's sort of related to how to deal with the layout stuff: 

![Screen Shot 2019-07-02 at 5 42 11 PM](https://user-images.githubusercontent.com/484309/60557540-c2795780-9cfa-11e9-85d8-130d082e64c1.png)

![Screen Shot 2019-07-02 at 5 42 27 PM](https://user-images.githubusercontent.com/484309/60557542-c4dbb180-9cfa-11e9-8f01-0e23f2a718be.png)

The redundant browse button still grates on me a bit, and I'd like to unify them. I think the main problem with the old browse button was just that it was invisible — I think we could get buy-in if the button was in the icon but was clearly visible. 

My current thinking is:

- When the form first loads, style the button like the above. Keep that styling at rest, so it's always visible.
- As soon as the user deliberately chooses an icon, drop the button styling so it stops obscuring their icon preview.
- But always show the caption + fade when you hover over it.
